curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderLatch.java

```diff
@@ -607,29 +607,38 @@ private void getChildren() throws Exception {
             public void processResult(CuratorFramework client, CuratorEvent event) throws Exception {
                 if (event.getResultCode() == KeeperException.Code.OK.intValue()) {
                     checkLeadership(event.getChildren());
+                } else if (event.getResultCode() == KeeperException.Code.NONODE.intValue()) {
+                    // latchPath has gone - reset
+                    //
+                    // This is possible when RECONNECTED during:
+                    // (1) Scale the zk cluster to 0 nodes.
+                    // (2) Scale it back.
+                    //
+                    // See also https://issues.apache.org/jira/browse/CURATOR-724
+                    reset();
+                } else {
+                    log.error("getChildren() failed (rc = {})", event.getResultCode());
                 }
             }
         };
         client.getChildren().inBackground(callback).forPath(ZKPaths.makePath(latchPath, null));
     }
 
+    @VisibleForTesting
+    volatile CountDownLatch debugHandleReconnectedLatch = null;
+
     @VisibleForTesting
     protected void handleStateChange(ConnectionState newState) {
         switch (newState) {
-            default: {
-                // NOP
-                break;
-            }
-
             case RECONNECTED: {
                 try {
-                    if (client.getConnectionStateErrorPolicy().isErrorState(ConnectionState.SUSPENDED)
-                            || !hasLeadership.get()) {
-                        getChildren();
+                    if (debugHandleReconnectedLatch != null) {
+                        debugHandleReconnectedLatch.await();
                     }
+                    getChildren();
```

Comment:
Also it doesn't hurt we always call `getChildren` to check our leadership. It's not reset now anyway.